### ISSUE TITLE
Depth anything V2 Tflite outputs constants on qualcomm gpus

### ISSUE THREAD

**Issue by at-brinc** [tensorflow/tensorflow#93476]
### 1. System information

- Ubuntu 20.04
- bulit grom source tensorflow and tflite 2.17.1
- v2.17.1 tensorflow 3c92ac03cab816044f7b18a86eb86aa01a294d95

### 2. Code

Provide code to help us reproduce your issues using one of the following options:
tflite: https://huggingface.co/qualcomm/Depth-Anything-V2/tree/main

pip3 install qai-hub
qai-hub configure --api_token INSERT_API_TOKEN

import qai_hub as hub
import numpy as np

# TFlite model path
tflite_model_id = "tflite_path"

# Setup input data
input_tensor = np.arange(np.prod((1, 518, 518, 3)), dtype=np.float32).reshape((1, 518, 518, 3))

# Submit inference job
job = hub.submit_inference_job(
    model=tflite_model_id,
    device=hub.Device("QCS8550 (Proxy)"),
    name="depth_anything_v2",
    inputs=dict(image=[input_tensor]),
    options="--compute_unit gpu"
)

# Load the output data into a dictionary of numpy arrays
output_tensors = job.download_output_data()

### 3. Failure after conversion
If the conversion is successful, but the generated model is wrong, then state what is wrong:

Model produces wrong results on qualcomm gpus. In qai-hub I replaced the GELU layers in the huggingface config to gelu_new and the model seems to export just fine using qai-hub. When I run it on the QRB5165 it seems to run fine on the gpu but outputs a constant image with value 0.088684. This model works fine if I run it on the cpu though. Also other models are working on the gpu if I use them. Also it seems like this model works completely fine on gpu delegate using opencl on cuda hardware on my laptop. It seems like if the model works on cpu the tflite is correct and hence how the graph is executed depends on gpu delegates.

**Comment by at-brinc** [comment]
Any updates here?

**Comment by at-brinc** [comment]
@gaikwadrahul8 Any updates

**Comment by at-brinc** [comment]
Any updates?

**Comment by gaikwadrahul8** [comment]
Hi, @at-brinc
I apologize for the delay in my response, As far I know the primary issue is that the GELU activation function is not supported by the TensorFlow Lite GPU delegate which is critical for the Depth Anything V2 model. When unsupported operations are encountered the GPU delegate fails to execute these operations resulting in constant output values (0.088684 in this case)

You can verify GELU operations in model using something like below 
```
import tensorflow as tf

interpreter = tf.lite.Interpreter(model_path="depth_anything_v2.tflite")
all_ops = list(map(lambda x: x["op_name"], interpreter._get_ops_details()))
print("Operations in model:", np.unique(all_ops))
```
You can verify which operations are being delegated to GPU vs CPU please refer this [LiteRT Model Analyzer](https://ai.google.dev/edge/litert/models/model_analyzer) and also please refer [GPU ML operations support](https://ai.google.dev/edge/litert/performance/gpu#supported_ops)
```
# Enable verbose logging to see delegation warnings
tf.lite.experimental.Analyzer.analyze(model_content=tflite_model, gpu_compatibility=True)
```
If I have missed something here please let me know. Thank you for your cooperation and patience.

**Comment by at-brinc** [comment]
Hi @gaikwadrahul8 
Can you explain why then it works on gpu delegate on nvidia hardware.

I am using below code to run it on gpu:
`
#include <iostream>
#include <chrono>
#include <opencv2/opencv.hpp>
#include "tensorflow/lite/interpreter.h"
#include "tensorflow/lite/kernels/register.h"
#include "tensorflow/lite/model.h"
#include "tensorflow/lite/delegates/gpu/delegate.h"
#include "tensorflow/lite/delegates/gpu/delegate_options.h"

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <model_path> <image_path>" << std::endl;
        return -1;
    }
    
    std::string model_path = argv[1];
    std::string image_path = argv[2];
    
    // Load input image
    cv::Mat input_image = cv::imread(image_path);
    if (input_image.empty()) {
        std::cerr << "Failed to load image: " << image_path << std::endl;
        return -1;
    }
    
    // Load model
    auto model = tflite::FlatBufferModel::BuildFromFile(model_path.c_str());
    if (!model) {
        std::cerr << "Failed to load model" << std::endl;
        return -1;
    }
    
    // Build interpreter
    tflite::ops::builtin::BuiltinOpResolver resolver;
    tflite::InterpreterBuilder builder(*model, resolver);
    std::unique_ptr<tflite::Interpreter> interpreter;
    builder(&interpreter);
    
    if (!interpreter) {
        std::cerr << "Failed to build interpreter" << std::endl;
        return -1;
    }
    
    // Setup GPU delegate
    TfLiteGpuDelegateOptionsV2 options = TfLiteGpuDelegateOptionsV2Default();
    auto* gpu_delegate = TfLiteGpuDelegateV2Create(&options);
    
    if (gpu_delegate && interpreter->ModifyGraphWithDelegate(gpu_delegate) == kTfLiteOk) {
        std::cout << "GPU delegate applied successfully" << std::endl;
    } else {
        std::cout << "GPU delegate failed, using CPU" << std::endl;
        if (gpu_delegate) {
            TfLiteGpuDelegateV2Delete(gpu_delegate);
            gpu_delegate = nullptr;
        }
    }
    
    // Allocate tensors
    if (interpreter->AllocateTensors() != kTfLiteOk) {
        std::cerr << "Failed to allocate tensors" << std::endl;
        return -1;
    }
    
    // Get input tensor info
    int input_idx = interpreter->inputs()[0];
    TfLiteTensor* input_tensor = interpreter->tensor(input_idx);
    
    // Preprocess image
    cv::Mat resized_image;
    cv::Size target_size(input_tensor->dims->data[2], input_tensor->dims->data[1]);
    cv::resize(input_image, resized_image, target_size);
    
    cv::Mat float_image;
    resized_image.convertTo(float_image, CV_32F, 1.0/255.0);
    
    // Copy data to input tensor
    float* input_data = interpreter->typed_input_tensor<float>(0);
    std::memcpy(input_data, float_image.data, float_image.total() * float_image.elemSize());
    
    // Run inference
    auto start = std::chrono::high_resolution_clock::now();
    if (interpreter->Invoke() != kTfLiteOk) {
        std::cerr << "Inference failed" << std::endl;
        return -1;
    }
    auto end = std::chrono::high_resolution_clock::now();
    
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::cout << "Inference time: " << duration.count() << " ms" << std::endl;
    
    // Get output
    int output_idx = interpreter->outputs()[0];
    TfLiteTensor* output_tensor = interpreter->tensor(output_idx);
    float* output_data = interpreter->typed_output_tensor<float>(0);
    
    // Create output image
    int height = output_tensor->dims->data[1];
    int width = output_tensor->dims->data[2];
    cv::Mat output_image(height, width, CV_32F, output_data);
    
    // Normalize and save
    cv::Mat normalized_output;
    cv::normalize(output_image, normalized_output, 0, 255, cv::NORM_MINMAX, CV_8U);
    cv::imwrite("output.png", normalized_output);
    
    std::cout << "Output saved to output.png" << std::endl;
    
    // Cleanup
    if (gpu_delegate) {
        TfLiteGpuDelegateV2Delete(gpu_delegate);
    }
    
    return 0;
}
`

I am getting:
`INFO: Created TensorFlow Lite delegate for GPU.
INFO: Loaded OpenCL library with dlopen.
INFO: Initialized OpenCL-based API.
INFO: Created 1 GPU delegate kernels.
GPU delegate applied successfully
Inference time: 90 ms
Output saved to output.png`

The output is correct and is not constant.

<img width="518" height="518" alt="Image" src="https://github.com/user-attachments/assets/99c4d9a2-a7e1-496a-add6-81b8434376d4" />

**Comment by at-brinc** [comment]
Anything here

**Comment by at-brinc** [comment]
Hi @gaikwadrahul8 . I was able to debug this. It seems like if I use fp16 the kernels are overflowing. Are all the kernels in tflite only implemented for fp32 or is there a way to tell tflite at runtime to use fp16 specific kernel which can keep track of the values in calculations and make sure that high values are either clamped or the fp32 version of that kernel is used.

**Comment by at-brinc** [comment]
It seems like padding in images causes issues in softmax kernel. PR: #104247

**Comment by nileshau-afk** [comment]
Hi @at-brinc,

Thank you for your patience, and apologies for the late reply. The issue where your Depth Anything V2 TFLite model outputs constant values on Qualcomm GPUs via `qai-hub` while working correctly on CPU and other GPU delegates indicates a problem with the Qualcomm GPU delegate's execution of certain operations or its precision handling. Since the model works on CPU, the TFLite conversion itself is likely correct, pointing to a delegate-specific behavior.

To diagnose this further, please try enabling verbose logging for the TFLite runtime and the GPU delegate. This can often reveal unsupported operations or warnings during graph preparation and execution. While `qai-hub` abstracts some of this, there might be ways to pass through TFLite interpreter options.

You can also review the general TensorFlow Lite GPU delegate documentation for common issues and configuration options: https://www.tensorflow.org/lite/performance/gpu

If you can provide the verbose logs from the Qualcomm GPU delegate execution, it would help pinpoint the exact cause.

Please test if this helps in diagnosing the issue.
